File: routes/simulation.py
"""
Simulation routes
"""

from flask import Blueprint, render_template, jsonify, request, session
from database import db
from models import Simulation, Resource, Territory
from datetime import datetime
import json
import random

bp = Blueprint('simulation', __name__, url_prefix='/simulation')

@bp.route('/')
def index():
    """Simulation mode interface"""
    return render_template('simulation/index.html')

@bp.route('/start', methods=['POST'])
def start_simulation():
    """Start a new simulation"""
    if 'user_id' not in session:
        return jsonify({'error': 'Not authenticated'}), 401
    
    data = request.json
    scenario_name = data.get('scenario_name', 'Default Scenario')
    start_year = data.get('start_year', 1941)
    side = data.get('side', 'allies')
    
    simulation = Simulation(
        user_id=session['user_id'],
        scenario_name=scenario_name,
        start_year=start_year,
        side=side,
        decisions=json.dumps([]),
        outcome=json.dumps({})
    )
    db.session.add(simulation)
    db.session.commit()
    
    return jsonify({
        'simulation_id': simulation.id,
        'scenario_name': scenario_name,
        'start_year': start_year,
        'side': side
    })

@bp.route('/decision', methods=['POST'])
def make_decision():
    """Process a strategic decision"""
    if 'user_id' not in session:
        return jsonify({'error': 'Not authenticated'}), 401
    
    data = request.json
    simulation_id = data.get('simulation_id')
    decision_type = data.get('decision_type')  # 'resource_allocation', 'espionage', 'military_action', 'diplomacy'
    decision_data = data.get('decision_data', {})
    
    simulation = Simulation.query.get(simulation_id)
    if not simulation or simulation.user_id != session['user_id']:
        return jsonify({'error': 'Simulation not found'}), 404
    
    # Load existing decisions
    decisions = json.loads(simulation.decisions or '[]')
    
    # Process decision and generate outcome
    outcome = evaluate_decision(decision_type, decision_data, simulation)
    
    # Add new decision
    decisions.append({
        'type': decision_type,
        'data': decision_data,
        'outcome': outcome,
        'timestamp': datetime.utcnow().isoformat()
    })
    
    # Update simulation
    simulation.decisions = json.dumps(decisions)
    simulation.outcome = json.dumps(outcome)
    db.session.commit()
    
    return jsonify({
        'success': True,
        'outcome': outcome,
        'decisions': decisions
    })

def evaluate_decision(decision_type, decision_data, simulation):
    """Evaluate a decision and return outcome"""
    # Simple probability-based evaluation
    base_success = 0.5
    
    if decision_type == 'resource_allocation':
        # Resource allocation decisions
        resource = decision_data.get('resource')
        amount = decision_data.get('amount', 0)
        target = decision_data.get('target', 'production')
        
        success_rate = min(0.9, base_success + (amount / 1000) * 0.1)
        success = random.random() < success_rate
        
        return {
            'success': success,
            'message': f"Resource allocation {'succeeded' if success else 'failed'}",
            'impact': {
                'production': amount * 0.1 if success else 0,
                'morale': 5 if success else -2
            }
        }
    
    elif decision_type == 'espionage':
        # Espionage missions
        target = decision_data.get('target')
        mission_type = decision_data.get('mission_type', 'intelligence')
        
        success_rate = 0.4  # Espionage is risky
        success = random.random() < success_rate
        
        return {
            'success': success,
            'message': f"Espionage mission {'succeeded' if success else 'was discovered'}",
            'intelligence_gained': random.randint(10, 50) if success else 0,
            'casualties': random.randint(0, 5) if not success else 0
        }
    
    elif decision_type == 'military_action':
        # Military operations
        operation_type = decision_data.get('operation_type')
        location = decision_data.get('location')
        forces = decision_data.get('forces', 1000)
        
        success_rate = min(0.8, base_success + (forces / 10000) * 0.2)
        success = random.random() < success_rate
        
        casualties = int(forces * random.uniform(0.1, 0.3))
        
        return {
            'success': success,
            'message': f"Military action {'succeeded' if success else 'failed'}",
            'casualties': casualties,
            'territory_gained': 1 if success else 0,
            'morale_impact': 10 if success else -5
        }
    
    elif decision_type == 'diplomacy':
        # Diplomatic actions
        action = decision_data.get('action')
        target_nation = decision_data.get('target_nation')
        
        success_rate = 0.6
        success = random.random() < success_rate
        
        return {
            'success': success,
            'message': f"Diplomatic action {'succeeded' if success else 'failed'}",
            'alliance_strength': 5 if success else -2,
            'resources_gained': random.randint(100, 500) if success else 0
        }
    
    return {
        'success': False,
        'message': 'Unknown decision type'
    }

@bp.route('/<int:simulation_id>')
def view_simulation(simulation_id):
    """View a specific simulation"""
    simulation = Simulation.query.get(simulation_id)
    if not simulation:
        return redirect(url_for('simulation.index'))
    
    return render_template('simulation/view.html', simulation=simulation)

File: static/js/simulation.js
// Simulation JavaScript

let currentSimulation = null;

// Initialize simulation
document.addEventListener('DOMContentLoaded', function() {
    const form = document.getElementById('scenario-form');
    form.addEventListener('submit', startSimulation);
});

// Start simulation
function startSimulation(e) {
    e.preventDefault();
    
    const scenarioName = document.getElementById('scenario-name').value;
    const startYear = document.getElementById('start-year').value;
    const side = document.getElementById('side').value;
    
    fetch('/simulation/start', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            scenario_name: scenarioName,
            start_year: parseInt(startYear),
            side: side
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.error) {
            alert('Error: ' + data.error);
            return;
        }
        
        currentSimulation = data;
        document.getElementById('decision-panel').style.display = 'block';
        
        displayOutcome({
            success: true,
            message: `Simulation "${scenarioName}" started. Year: ${startYear}, Side: ${side.toUpperCase()}`,
            simulation_id: data.simulation_id
        });
    })
    .catch(error => {
        console.error('Error starting simulation:', error);
        alert('Failed to start simulation');
    });
}

// Show decision form
function showDecisionForm(decisionType) {
    const container = document.getElementById('decision-form-container');
    
    let formHTML = '';
    
    switch(decisionType) {
        case 'resource_allocation':
            formHTML = `
                <div class="simulation-form">
                    <div class="form-group">
                        <label>RESOURCE TYPE</label>
                        <select id="resource-type" class="form-input">
                            <option value="oil">Oil</option>
                            <option value="steel">Steel</option>
                            <option value="manpower">Manpower</option>
                        </select>
                    </div>
                    <div class="form-group">
                        <label>AMOUNT</label>
                        <input type="number" id="resource-amount" class="form-input" value="1000" min="100" step="100">
                    </div>
                    <div class="form-group">
                        <label>TARGET</label>
                        <select id="resource-target" class="form-input">
                            <option value="production">Production</option>
                            <option value="military">Military</option>
                            <option value="research">Research</option>
                        </select>
                    </div>
                    <button class="btn-military btn-primary" onclick="makeDecision('resource_allocation')">
                        EXECUTE
                    </button>
                </div>
            `;
            break;
            
        case 'espionage':
            formHTML = `
                <div class="simulation-form">
                    <div class="form-group">
                        <label>TARGET NATION</label>
                        <select id="espionage-target" class="form-input">
                            <option value="Germany">Germany</option>
                            <option value="Japan">Japan</option>
                            <option value="Italy">Italy</option>
                            <option value="USSR">USSR</option>
                            <option value="USA">USA</option>
                        </select>
                    </div>
                    <div class="form-group">
                        <label>MISSION TYPE</label>
                        <select id="espionage-type" class="form-input">
                            <option value="intelligence">Intelligence Gathering</option>
                            <option value="sabotage">Sabotage</option>
                            <option value="propaganda">Propaganda</option>
                        </select>
                    </div>
                    <button class="btn-military btn-primary" onclick="makeDecision('espionage')">
                        LAUNCH MISSION
                    </button>
                </div>
            `;
            break;
            
        case 'military_action':
            formHTML = `
                <div class="simulation-form">
                    <div class="form-group">
                        <label>OPERATION TYPE</label>
                        <select id="operation-type" class="form-input">
                            <option value="offensive">Offensive</option>
                            <option value="defensive">Defensive</option>
                            <option value="raid">Raid</option>
                        </select>
                    </div>
                    <div class="form-group">
                        <label>LOCATION</label>
                        <input type="text" id="operation-location" class="form-input" placeholder="Enter location">
                    </div>
                    <div class="form-group">
                        <label>FORCES</label>
                        <input type="number" id="operation-forces" class="form-input" value="5000" min="1000" step="1000">
                    </div>
                    <button class="btn-military btn-primary" onclick="makeDecision('military_action')">
                        EXECUTE OPERATION
                    </button>
                </div>
            `;
            break;
            
        case 'diplomacy':
            formHTML = `
                <div class="simulation-form">
                    <div class="form-group">
                        <label>ACTION</label>
                        <select id="diplomacy-action" class="form-input">
                            <option value="negotiate">Negotiate Treaty</option>
                            <option value="alliance">Form Alliance</option>
                            <option value="trade">Trade Agreement</option>
                        </select>
                    </div>
                    <div class="form-group">
                        <label>TARGET NATION</label>
                        <select id="diplomacy-target" class="form-input">
                            <option value="USA">USA</option>
                            <option value="USSR">USSR</option>
                            <option value="United Kingdom">United Kingdom</option>
                            <option value="Germany">Germany</option>
                            <option value="Japan">Japan</option>
                        </select>
                    </div>
                    <button class="btn-military btn-primary" onclick="makeDecision('diplomacy')">
                        INITIATE DIPLOMACY
                    </button>
                </div>
            `;
            break;
    }
    
    container.innerHTML = formHTML;
}

// Make decision
function makeDecision(decisionType) {
    if (!currentSimulation) {
        alert('Please start a simulation first');
        return;
    }
    
    let decisionData = {};
    
    switch(decisionType) {
        case 'resource_allocation':
            decisionData = {
                resource: document.getElementById('resource-type').value,
                amount: parseInt(document.getElementById('resource-amount').value),
                target: document.getElementById('resource-target').value
            };
            break;
        case 'espionage':
            decisionData = {
                target: document.getElementById('espionage-target').value,
                mission_type: document.getElementById('espionage-type').value
            };
            break;
        case 'military_action':
            decisionData = {
                operation_type: document.getElementById('operation-type').value,
                location: document.getElementById('operation-location').value,
                forces: parseInt(document.getElementById('operation-forces').value)
            };
            break;
        case 'diplomacy':
            decisionData = {
                action: document.getElementById('diplomacy-action').value,
                target_nation: document.getElementById('diplomacy-target').value
            };
            break;
    }
    
    fetch('/simulation/decision', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            simulation_id: currentSimulation.simulation_id,
            decision_type: decisionType,
            decision_data: decisionData
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.error) {
            alert('Error: ' + data.error);
            return;
        }
        
        displayOutcome(data.outcome);
    })
    .catch(error => {
        console.error('Error making decision:', error);
        alert('Failed to process decision');
    });
}

// Display outcome
function displayOutcome(outcome) {
    const container = document.getElementById('outcome-display');
    
    const outcomeClass = outcome.success ? 'outcome-success' : 'outcome-failure';
    const timestamp = new Date().toLocaleTimeString('en-US', {hour12: false});
    
    const outcomeEl = document.createElement('div');
    outcomeEl.className = `outcome-item ${outcomeClass}`;
    outcomeEl.innerHTML = `
        <div style="display: flex; justify-content: space-between; margin-bottom: 0.5rem">
            <strong>${outcome.success ? '✓ SUCCESS' : '✗ FAILURE'}</strong>
            <span style="font-size: 0.85rem; color: #999">${timestamp}</span>
        </div>
        <div style="margin-bottom: 0.5rem">${outcome.message}</div>
        ${outcome.impact ? `
            <div style="font-size: 0.85rem; margin-top: 0.5rem">
                <strong>IMPACT:</strong>
                ${Object.keys(outcome.impact).map(key => 
                    `${key}: ${outcome.impact[key] > 0 ? '+' : ''}${outcome.impact[key]}`
                ).join(', ')}
            </div>
        ` : ''}
        ${outcome.casualties !== undefined ? `
            <div style="font-size: 0.85rem; color: #ff0000; margin-top: 0.5rem">
                CASUALTIES: ${outcome.casualties}
            </div>
        ` : ''}
        ${outcome.intelligence_gained ? `
            <div style="font-size: 0.85rem; color: #CC6666; margin-top: 0.5rem">
                INTELLIGENCE GAINED: ${outcome.intelligence_gained}
            </div>
        ` : ''}
    `;
    
    container.insertBefore(outcomeEl, container.firstChild);
    
    // Keep only last 10 outcomes
    while (container.children.length > 10) {
        container.removeChild(container.lastChild);
    }
}

